### ISSUE TITLE
Timeout when calling Refresh()

### ISSUE THREAD

**Issue by rosrosros** [Azure/AppConfiguration#173]
I'm following this tutorial: https://docs.microsoft.com/en-us/azure/azure-app-configuration/enable-dynamic-configuration-dotnet

What am I doing wrong?

But doing multiple refresh calls I get this error:

```
Test 5
0, msg: Test 5, took: 59ms
1, msg: Test 5, took: 0ms
2, msg: Test 5, took: 0ms
3, msg: Test 5, took: 0ms
4, msg: Test 5, took: 0ms
5, msg: Test 5, took: 0ms
6, msg: Test 5, took: 0ms
7, msg: Test 5, took: 0ms
8, msg: Test 5, took: 0ms
9, msg: Test 5, took: 0ms
10, msg: Test 5, took: 125ms
11, msg: Test 5, took: 0ms
12, msg: Test 5, took: 0ms
13, msg: Test 5, took: 0ms
14, msg: Test 5, took: 0ms
15, msg: Test 5, took: 0ms
16, msg: Test 5, took: 0ms
17, msg: Test 5, took: 0ms
18, msg: Test 5, took: 0ms
19, msg: Test 5, took: 0ms

Unhandled Exception: System.AggregateException: One or more errors occurred. ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at AzureAppConfig.Program.Main(String[] args) in C:\Users\ros\Source\Repos\AzureAppConfig\AzureAppConfig\Program.cs:line 49
Press any key to continue . . .
```

My code:

`private static IConfigurationRefresher _configurationRefresher;

        static void Main(string[] args)
        {
            string message = System.Configuration.ConfigurationManager.AppSettings["TestApp:Settings:Message"];

            Console.WriteLine(message);
            //var connectionString = Environment.GetEnvironmentVariable("ConnectionString");

            var connectionString = System.Configuration.ConfigurationManager.AppSettings["ConnectionString"];
            //Console.WriteLine(connectionString);
            var configurationBuilder = new ConfigurationBuilder()
                .AddAzureAppConfiguration(options =>
                {
                    
                    options.Connect(connectionString);
                    options.ConfigureRefresh(refresh =>
                        refresh
                            .Register("TestApp:Settings:Message")
                            .SetCacheExpiration(TimeSpan.FromSeconds(10))
                    );

                    _configurationRefresher = options.GetRefresher();
                });

            IConfigurationRoot configuration = configurationBuilder.Build();
            for (int i = 0; i < 100; i++)
            {
                var startNew = Stopwatch.StartNew();
                
                _configurationRefresher.Refresh().Wait();
                var msg = configuration["TestApp:Settings:Message"];
                Console.WriteLine(i + ": " + msg + " " + startNew.ElapsedMilliseconds + "ms");

                Thread.Sleep(TimeSpan.FromSeconds(1));
            }`

Nuget packages:
```
<packages>
  <package id="Microsoft.AspNetCore.Http.Abstractions" version="2.1.1" targetFramework="net471" />
  <package id="Microsoft.AspNetCore.Http.Extensions" version="2.1.1" targetFramework="net471" />
  <package id="Microsoft.AspNetCore.Http.Features" version="2.1.1" targetFramework="net471" />
  <package id="Microsoft.Azure.KeyVault" version="3.0.3" targetFramework="net471" />
  <package id="Microsoft.Azure.KeyVault.WebKey" version="3.0.3" targetFramework="net471" />
  <package id="Microsoft.Azure.Management.ResourceManager" version="2.0.0-preview" targetFramework="net471" />
  <package id="Microsoft.Azure.Services.AppAuthentication" version="1.0.3" targetFramework="net471" />
  <package id="Microsoft.Configuration.ConfigurationBuilders.AzureAppConfiguration" version="1.0.0-beta" targetFramework="net471" />
  <package id="Microsoft.Configuration.ConfigurationBuilders.Base" version="2.0.0-beta" targetFramework="net471" />
  <package id="Microsoft.Extensions.Configuration" version="2.1.1" targetFramework="net471" />
  <package id="Microsoft.Extensions.Configuration.Abstractions" version="2.1.1" targetFramework="net471" />
  <package id="Microsoft.Extensions.Configuration.AzureAppConfiguration" version="2.0.0-preview-010050001-38" targetFramework="net471" />
  <package id="Microsoft.Extensions.FileProviders.Abstractions" version="2.1.1" targetFramework="net471" />
  <package id="Microsoft.Extensions.Primitives" version="2.1.1" targetFramework="net471" />
  <package id="Microsoft.IdentityModel.Clients.ActiveDirectory" version="3.14.2" targetFramework="net471" />
  <package id="Microsoft.Net.Http.Headers" version="2.1.1" targetFramework="net471" />
  <package id="Microsoft.Rest.ClientRuntime" version="2.3.18" targetFramework="net471" />
  <package id="Microsoft.Rest.ClientRuntime.Azure" version="3.3.18" targetFramework="net471" />
  <package id="Newtonsoft.Json" version="11.0.2" targetFramework="net471" />
  <package id="System.Buffers" version="4.5.0" targetFramework="net471" />
  <package id="System.Configuration.ConfigurationManager" version="4.7.0-preview1.19504.10" targetFramework="net471" />
  <package id="System.Interactive.Async" version="3.1.1" targetFramework="net471" />
  <package id="System.Memory" version="4.5.1" targetFramework="net471" />
  <package id="System.Numerics.Vectors" version="4.4.0" targetFramework="net471" />
  <package id="System.Reactive" version="4.0.0" targetFramework="net471" />
  <package id="System.Reactive.Linq" version="4.0.0" targetFramework="net471" />
  <package id="System.Runtime.CompilerServices.Unsafe" version="4.5.1" targetFramework="net471" />
  <package id="System.Security.AccessControl" version="4.7.0-preview1.19504.10" targetFramework="net471" />
  <package id="System.Security.Permissions" version="4.7.0-preview1.19504.10" targetFramework="net471" />
  <package id="System.Security.Principal.Windows" version="4.7.0-preview1.19504.10" targetFramework="net471" />
  <package id="System.Text.Encodings.Web" version="4.5.0" targetFramework="net471" />
</packages>

```

**Comment by abhilasharora** [comment]
@rosrosros To better understand the issue, could you share how many key-values exist in the configuration store that you're trying to access in your code?

**Comment by rosrosros** [comment]
Only a couple of test keys and feature flags.

**Comment by zhenlan** [comment]
@rosrosros thanks for reporting the issue. I can reproduce it with the code you provided. We will take a further look.

**Comment by abhilasharora** [comment]
@rosrosros We have a local fix for this issue and it would be included in the next release of the Nuget package `Microsoft.Extensions.Configuration.AzureAppConfiguration`.

**Comment by rosrosros** [comment]
@abhilasharora Thanks.
Is it correct that I should call `_configurationRefresher.Refresh().Wait()` before each call to `configuration["TestApp:Settings:Message"]` ?

Or do you have an example of another approach with using the dotnet framework?

**Comment by abhilasharora** [comment]
@rosrosros It is up to developers to invoke `_configurationRefresher.Refresh()` method as appropriate for their use case. In a typical case, the call could be made before each code snippet that expects a possibly updated value of the key-values registered for refresh. In case the code only expects to check an updated value in certain conditions, then the call could be made just before that conditional check. The example that you posted in the issue description is a valid usage of the method.

Please note that a call to `Refresh` may not actually fetch the updated value from the server if the cached value has not expired or if all attempts to fetch the value from the server fail.

![image](https://user-images.githubusercontent.com/47916289/67967932-b7103d00-fbc3-11e9-8e21-0528b22b0973.png)

**Comment by abhilasharora** [comment]
@rosrosros This issue has been fixed in the version `2.1.0-preview-010380001-1099` of the package `Microsoft.Extensions.Configuration.AzureAppConfiguration` that was released yesterday.
